Drop black dependency

akaihola opened the pull request:
This allows installing Darker without needing to install Black, and running e.g.
```bash
darker --formatter=none --isort
darker --formatter=none --flynt
```
and, once #744 has been merged, this will also work without needing to install Black:
```bash
darker --formatter=ruff
```

Before reviewing or merging this one:
- [x] Review #738
- [x] Merge #738
- [x] Rebase on `master`
- [x] Review #761
- [x] Merge #761
- [x] Rebase on `master`

clintonsteiner commented on src/darker/formatters/black_formatter.py:
```diff
@@ -0,0 +1,227 @@
+
+    def _read_config_file(self, config_path: str) -> None:  # noqa: C901
+        # Local import so Darker can be run without Black installed
+        from black import parse_pyproject_toml, re_compile_maybe_verbose
```
```suggestion
        from black import parse_pyproject_toml, re_compile_maybe_verbose # pylint: disable=import-outside-toplevel
```

clintonsteiner commented on src/darker/formatters/black_formatter.py:
```diff
@@ -0,0 +1,227 @@
+
+        """
+        # Local import so Darker can be run without Black installed
+        from black import format_str
```
```suggestion
        from black import format_str # pylint: disable=import-outside-toplevel
```

clintonsteiner commented on src/darker/formatters/black_formatter.py:
```diff
@@ -0,0 +1,227 @@
+        # since at this point we already have a single file's content to work on.
+
+        # Local import so Darker can be run without Black installed
+        from black import FileMode as Mode
```
```suggestion
        from black import FileMode as Mode # pylint: disable=import-outside-toplevel
```

clintonsteiner commented on src/darker/formatters/black_formatter.py:
```diff
@@ -0,0 +1,227 @@
+
+        # Local import so Darker can be run without Black installed
+        from black import FileMode as Mode
+        from black import TargetVersion
```
```suggestion
        from black import TargetVersion # pylint: disable=import-outside-toplevel
```

clintonsteiner commented on src/darker/tests/test_files.py:
```diff
@@ -1,22 +1,33 @@
-@patch("darker.files.find_user_pyproject_toml")
-def test_find_pyproject_toml(find_user_pyproject_toml: MagicMock) -> None:
+@pytest.mark.kwparametrize(
+    dict(start="only_pyproject/subdir", expect="only_pyproject/pyproject.toml"),
```
```suggestion
    {"start":"only_pyproject/subdir", "expect":"only_pyproject/pyproject.toml"},
```

akaihola replied:
This is a style decision I've made when using `pytest-kwparametrize`: to avoid a jungle of double quotes and make test cases more readable, I use the `dict()` constructor, and silence linter warnings in each module.

clintonsteiner commented on src/darker/tests/test_files.py:
```diff
@@ -1,22 +1,33 @@
-def test_find_pyproject_toml(find_user_pyproject_toml: MagicMock) -> None:
+@pytest.mark.kwparametrize(
+    dict(start="only_pyproject/subdir", expect="only_pyproject/pyproject.toml"),
+    dict(start="only_git/subdir", expect=None),
```
```suggestion
    {"start":"only_git/subdir", "expect":None},
```

clintonsteiner commented on src/darker/tests/test_files.py:
```diff
@@ -1,22 +1,33 @@
+@pytest.mark.kwparametrize(
+    dict(start="only_pyproject/subdir", expect="only_pyproject/pyproject.toml"),
+    dict(start="only_git/subdir", expect=None),
+    dict(start="git_and_pyproject/subdir", expect="git_and_pyproject/pyproject.toml"),
```
```suggestion
    {"start":"git_and_pyproject/subdir", "expect":"git_and_pyproject/pyproject.toml"},
```

clintonsteiner commented on src/darker/files.py:
```diff
@@ -1,27 +1,154 @@
+            root_relative_path = f"{root_relative_path}/"
+
+        if any(
+                _path_is_excluded(root_relative_path, x)
```
```suggestion
            _path_is_excluded(root_relative_path, x)
```

clintonsteiner commented on src/darker/files.py:
```diff
@@ -1,27 +1,154 @@
+
+        if any(
+                _path_is_excluded(root_relative_path, x)
+                for x in [exclude, extend_exclude, force_exclude]
```
```suggestion
            for x in [exclude, extend_exclude, force_exclude]
```

akaihola commented:
@clintonsteiner, I removed Black as a hard dependency, and added `darker[black]` extras instead. Documentation has also been updated accordingly. The default formatter is still Black, and a helpful error message is displayed if Black is needed but not installed.
